### ISSUE TITLE
taking mutability seriously

### ISSUE THREAD

**Issue by Roger-luo** [JuliaDiff/ChainRules.jl#382]
OK, I admit I can't find a better title for this. But I feel this should be brought up. See previous related issues in #381 #380 #46 

@GiggleLiu keeps wanting to make all the return type of `pullback` to be the same as its input, that's why he was asking how to fix the problem in #381 for all possible cases. However, I don't think this would work with the current `thunk` mechanism. 

I'm thinking there should be a mutability trait to handle this task, and a BangBang version of `setindex!` (which is `setindex!!`) so when the rules writer requires the interface, ChainRules can check and error. 

Meanwhile, I think part of the problem of using `thunk` is because current AD compilers like Zygote are not able to do optimization if the compiler is able to do what `thunk` does, e.g Yota can figure out the gradient accumulation and unused gradient I believe, then it is not necessary to do it lazily. 

cc: @oxinabox @sethaxen

**Comment by oxinabox** [comment]
AFAICT this has nothing to do with the thunking mechanism.

**Comment by Roger-luo** [comment]
> AFAICT this has nothing to do with the thunking mechanism.

I'm just saying one cannot make the input and output type of a pullback to be the same, or in this case, creating `thunk` will be disallowed. which is what @GiggleLiu proposed

**Comment by GiggleLiu** [comment]
@Roger-luo thanks for openning an issue, this is an important topic that deserves extensive discussion. I wish someone can convince me that why output type can not be same as adjoint. I am unimpressed by `thunk`, because pytorch can use a single type `tensor` to get very good performance. To me, it seems "returning the same type or nothing" can do most of the job, this will make it so much easier for debugging.
